Update multiconfig package approach N configs -> 1 package (#1322)
conan-io/docs: creating_packages/package_approaches.rst

.. note::

    A complete working example of the following code can be found in the examples repo: https://github.com/conan-io/examples

    .. code:: bash

        $ git clone https://github.com/conan-io/examples.git
        $ cd features/multi_config
        $ conan create . user/channel

Creating a multi-configuration debug/release package is simple

The first step will be to remove ``build_type`` from the settings. It will not be an input setting and the generated package will always

> contain both debug and release artifacts.

Reviewer: may we add some clarification, what happens to other configurations besides Release/Debug (e.g. RelWithDebInfo), is it unsupported case or it will just work out of the box?

Author: probably it will just work, but I don't want to make the example more complex or make the explanation too long for the user

Reviewer 2: I think it is ok for this PR to keep the Debug/Release as it was before.

The Visual Studio runtime is different for debug and release (``MDd`` or ``MD``) and is set using the default runtime (``MD``/``MDd``). If
this meets your needs, we recommend removing the ``compiler.runtime`` subsetting in the ``configure()`` method:

.. code-block:: python

    class HelloConan(ConanFile):
        # build_type has been omitted. It is not an input setting.
        settings = "os", "compiler", "arch"
        generators = "cmake"

        # Remove runtime and use always default (MD/MDd)
        def configure(self):
            if self.settings.compiler == "Visual Studio":
                del self.settings.compiler.runtime

        def build(self):

>             cmake_release = CMake(self, build_type="Debug")

Reviewer: doesn't it require different build folders to be used? otherwise, I guess Release will just override Debug (or fail?)

Author: name of the generated artifacts are different, check the CMakeLists of the example

            cmake_release.configure()
            cmake_release.build()

            cmake_debug = CMake(self, build_type="Release")
            cmake_debug.configure()
            cmake_debug.build()

> In this case, we assume that the binaries will be differentiated with a suffix in the CMake syntax:

Reviewer 2: I think it would be more clear if we provide more information:

```suggestion
In this example, the binaries will be differentiated with a suffix in the CMake syntax, so we have to add this information to the data provided to the consumers in the `package_info` function:
```

and in the code-block below add the full function:

```python
def package_info(self):
    self.cpp_info.release.libs = ["mylibrary"]
    self.cpp_info.debug.libs = ["mylibrary_d"]
```

        self.cpp_info.release.libs = ["mylibrary"]
        self.cpp_info.debug.libs = ["mylibrary_d"]

> This will translate to the CMAake variables:

Reviewer: ```suggestion
This will translate to the CMake variables:
```

    set(CONAN_LIBS_DEBUG mylibrary_d ${CONAN_LIBS_DEBUG})
    set(CONAN_LIBS_RELEASE mylibrary ${CONAN_LIBS_RELEASE})

> And these variables will be correctly applied to each configuration by ``conan_basic_setup()`` helper.

Reviewer 2: What about other generators? Maybe it is useful to mention the `cmake_find_package` at least as it is the one we are moving to. I suppose the target created by Conan in the `FindXXX.cmake` uses the proper library depending on the build_type, right?

Author: Here I just fixed the outdated information to adequate it to the example, I didn't want to introduce in this PR new information.

I am sure the whole docs would benefit from a review regarding the cmake generators, but we still have to clear the path towards the full usability of the find_package generator.

I will try what you suggest and open a new issue accordingly.

Reviewer: I think the only generators currently supported are `cmake` and `visual_studio` (plus their corresponding `multi` variants). we should mention that. 
besides that, what happens to `conan_basic_setup(TARGETS)`, is it supported as well?

Reviewer 2: I agree it deserves an investigation to know if it works or not, but it is not the aim of this PR. @danimtb , if you open the new issue related to this example, please add the concerns by @SSE4 so we have a look into it too.

.. important::

    The above code assumes that the package will always use the default Visual Studio runtime (``MD``/``MDd``).
    To keep the package configurable for supporting static(MT)/dynamic(MD) linking with the VS runtime
    library, you can do the following:

    - Keep the ``compiler.runtime`` setting, e.g. do not implement the ``configure()`` method removing it.
    - Don't let the CMake helper define the ``CONAN_LINK_RUNTIME`` variable to define the runtime and define ``CONAN_LINK_RUNTIME_MULTI``
      instead.
    - In *CMakeLists.txt*, use the ``CONAN_LINK_RUNTIME_MULTI`` variable to correctly setup up the runtime for debug and release flags.
    - Write a separate ``package_id()`` methods for MD/MDd and for MT/MTd defining the packages to be built.

    All these steps are already coded in the repo https://github.com/conan-io/examples/tree/master/features/multi_config and commented
    out as **"Alternative 2"**.

Also, you can use any custom configuration as they are not restricted. For example, if your package is a multi-library package, you could

> try doing something like:

Reviewer 2: ```suggestion
try to do something like:
```

.. code-block:: python

    def package_info(self):
        self.cpp_info.regex.libs = ["myregexlib1", "myregexlib2"]
        self.cpp_info.filesystem.libs = ["myfilesystemlib"]

These specific config variables will not be automatically applied, but you can directly use them in your consumer CMake build script.

.. note::

>     The automatic conversion of multi-config variables to generators is currently only implemented in the ``cmake`` and ``txt`` generators.

Reviewer: I think `visual_studio` & `visual_studio_multi` generators also support it for a while

Author: You are right. However, I am not sure if adding the ``*_multi`` generators here would be confusing, as those are used only for consumption. Maybe @memsharded can give us a hint here

>     The automatic conversion of multi-config variables to generators is currently only implemented in the ``cmake`` and ``txt`` generators.

Author: ```suggestion
    The automatic conversion of multi-config variables to generators is currently only implemented in the ``cmake``, ``visual_studio`` and ``txt`` generators.
```

    If you want to have support for them in another build system, please open a GitHub issue.

N configs (1 build) -> N packages
---------------------------------

It’s possible that an existing build script is simultaneously building binaries for different
configurations, like debug/release, or different architectures (32/64bits), or library types
(shared/static). If such a build script is used in the previous "Single configuration packages"
approach, it will definitely work without problems. However, we'll be wasting precious build time, as

> we’ll be re-building the rebuilding project for each package, then extracting the relevant artifacts for

Reviewer 2: ```suggestion
we’ll be re-building the project for each package, then extracting the relevant artifacts for
```
